### DISCUSSION TITLE
Intercept - Modify HTTP Response Headers

### DISCUSSION THREAD

**Discussion by marktamis** [getgauge/taiko#1745] [General]
Hi,

We're looking to intercept HTTP response headers to rewrite them before the page is shown in the browser.

For my test setup we want to load a script from an external site into pages protected by restrictive Content Security Policy headers

Currently we can add the script but because of the aforementioned setting, it won't load

The idea would be to intercept and rewrite this response header to add  the domain from which my script is loading

(normally the script is loaded through a browser extension - currently we can force the install but we're stuck on user validation of the install, which means we can't use headless chrome)

**Comment by zabil** [comment]
Have you tried disabling web security? https://docs.taiko.dev/frequently_asked_questions/#how-do-i-intercept-request-with-method-options%3F

**Reply by marktamis** [reply]
Thanks! I just tried it but unfortunately it has no effect on the CSP Header (only CORS _requests_ are affected and no effect on the _response headers_)

**Comment by NivedhaSenthil** [comment]
@marktamis is there a sample that we can experiment with ?

**Reply by marktamis** [reply]
My objective to test an online guidance platform overlay running on an Enterprise application. Currently its libraries are injected into the application  by an extension. The role of the extension is to modify the CSP headers so that scripts hosted on a remote server are allowed to be added to the page (we have no way to modify the application's CSP headers directly).

As we can't add an extension with Taiko in headless mode, I want to intercept the response and modify CSP header, and then run an evaluate step to inject the script

I have set up a copy of the taiko todo app and added a restrictive csp header. We could either scrape out the nonce, or remove it and add an exception for the remote server - or remove the csp header alltogether when running the test

The copy can be accessed here: https://todo.touchflows.com

```
// nonce value should be scaped from the csp header and stored f.e. on the window object, in a cookie or sessionStorage
content-security-policy: default-src 'self'; script-src 'self' 'nonce-7fb1829a3065fdfac725695b0ae9fc3c'; style-src 'self' https: 'unsafe-inline';img-src 'self' http: https: data: blob: *.netlify.com

function extractNonce(csp) {
  var rx =  #/\w[a-f0-9]{31}/g;
  var arr = rx.exec(csp);
  return arr[0];
}
window.nonceValue = extractNonce(csp)
```


Adding the script during the test is done through evaluate:

**Inject snippet test step**
```
step("Inject snippet", async function() 
{
  await evaluate(() => { 
      var snippet= document.createElement('script'); 
           snippet.type = 'text/javascript'; 
           snippet.async = true; 
           //snippet.nonce= window.nonceValue; 
           snippet.src = 'https://marktamis.com/wp-content/uploads/2020/12/loader.js'; 
           var s = document.getElementsByTagName('script')[0]; 
           s.parentNode.insertBefore(snippet, s); 
        });
})
```
the loader script will then inject popper.js & tippy.js in order to simulate the online guidance system to add two tooltips to the todo app

**loader.js**

```
var s = document.getElementsByTagName('script')[0];


var popperjs = document.createElement('script'); 
popperjs.type = 'text/javascript'; 
popperjs.nonce = nonceValue;
popperjs.async = true; 
popperjs.src = 'https://unpkg.com/@popperjs/core@2'; 
s.parentNode.insertBefore(popperjs, s);

var tippy = document.createElement('script'); 
tippy.type = 'text/javascript'; 
tippy.nonce = nonceValue;
tippy.async = true; 
tippy.src = 'https://unpkg.com/tippy.js@6'; 
s.parentNode.insertBefore(tippy, s); 

var csslink = document.createElement('link'); 
csslink.rel = 'stylesheet';
csslink.type= 'text/css';
csslink.href = 'https://unpkg.com/tippy.js@6/animations/scale.css'; 
var l = document.getElementsByTagName('link')[0];
l.parentNode.insertBefore(csslink, l);

window.onload = function () {
	tippy('.todo-list', {
			content: 'Mark as completed by clicking the circle in front of the task',
	});
	tippy('.clear-completed', {
			content: 'Clear all completed tasks',
	});
}
```
Note the use of the nonce value. you could extract the nonce once you have the csp header with a regex

**Reply by sriv** [reply]
I believe this should be an issue (apologies for the late response), so that this gets slotted in the backlog.

**Reply by sriv** [reply]
https://github.com/getgauge/taiko/issues/1822 could be related to this.

**Reply by marktamis** [reply]
Thanks for picking this up again

When searching around I found that on other testing frameworks they add a method to remove the CSP header : bypassCSP (boolean)

This could be another way Taiko could handle the requirement

**Reply by nseshadr** [reply]
is there an ETA for this solution? This is critical for us to use Taiko.

**Reply by NivedhaSenthil** [reply]
Have created an issue to provide option for bypassing csp header https://github.com/getgauge/taiko/issues/1935
